### PR TITLE
Fix CEL cost handling of zero length replacement strings

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes issues found in ppc64le tests of CEL estimated cost for the `replace()` function: https://github.com/kubernetes/kubernetes/pull/119800#issuecomment-1681945591

The difference in cost was caused by ppc64le calculating this expression differently than amd64:

```go
uint64(math.Ceil(float64(sz.Max)/float64(toReplaceSz.Min))) * replaceWithSz.Max
```

https://gist.github.com/jpbetz/032b54ce4ef8f5427c4812c023f6c693 shows the exact difference, which is that `uint64(+Inf)`  results in 9223372036854775808 for amd64 but 18446744073709551615 for ppc64le (and arm64?).

The fix is to avoid the divide the creates the +Inf by handling the cost edge cases for `replace()`.  I've verified this fix works on both architectures.

#### Does this PR introduce a user-facing change?

```release-note
Fixed CEL estimated cost of `replace()` to handle a zero length replacement string correctly.
Previously this would cause the estimated cost to be higher than it should be.
```

/sig api-machinery

### CODE REVIEW

**Comment by Rajalakshmi-Girish** [comment]
/lgtm
It fixes the failure on `ppc64le`
```
[root@raji-workspace2 kubernetes]# uname -m
ppc64le
[root@raji-workspace2 kubernetes]#
[root@raji-workspace2 fix]# git clone https://github.com/jpbetz/kubernetes
Cloning into 'kubernetes'...
remote: Enumerating objects: 1418172, done.
remote: Counting objects: 100% (201/201), done.
remote: Compressing objects: 100% (111/111), done.
remote: Total 1418172 (delta 101), reused 123 (delta 85), pack-reused 1417971
Receiving objects: 100% (1418172/1418172), 979.50 MiB | 16.09 MiB/s, done.
Resolving deltas: 100% (1040719/1040719), done.
Updating files: 100% (23557/23557), done.
[root@raji-workspace2 fix]# cd kubernetes/
[root@raji-workspace2 kubernetes]# go test -v -timeout 30s -run ^TestCostEstimation/extended_library_replace$ k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel
=== RUN   TestCostEstimation
=== RUN   TestCostEstimation/extended_library_replace
=== RUN   TestCostEstimation/extended_library_replace/calc_maxLength
=== RUN   TestCostEstimation/extended_library_replace/set_maxLength
--- PASS: TestCostEstimation (0.01s)
    --- PASS: TestCostEstimation/extended_library_replace (0.01s)
        --- PASS: TestCostEstimation/extended_library_replace/calc_maxLength (0.01s)
        --- PASS: TestCostEstimation/extended_library_replace/set_maxLength (0.00s)
PASS
ok      k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel 0.022s
[root@raji-workspace2 kubernetes]# git checkout fix-ppc64le
Updating files: 100% (15550/15550), done.
branch 'fix-ppc64le' set up to track 'origin/fix-ppc64le'.
Switched to a new branch 'fix-ppc64le'
[root@raji-workspace2 kubernetes]# go test -v -timeout 30s -run ^TestCostEstimation/extended_library_replace$ k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel
=== RUN   TestCostEstimation
=== RUN   TestCostEstimation/extended_library_replace
=== RUN   TestCostEstimation/extended_library_replace/calc_maxLength
=== RUN   TestCostEstimation/extended_library_replace/set_maxLength
--- PASS: TestCostEstimation (0.01s)
    --- PASS: TestCostEstimation/extended_library_replace (0.01s)
        --- PASS: TestCostEstimation/extended_library_replace/calc_maxLength (0.01s)
        --- PASS: TestCostEstimation/extended_library_replace/set_maxLength (0.00s)
PASS
ok      k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel 0.032s
[root@raji-workspace2 kubernetes]#
```

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 1]
```diff
@@ -126,13 +126,32 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
...
+			var perReplaceWithSz, retainedSz checker.SizeEstimate
+			if toReplaceSz.Min == 0 {
+				// if the string being replaced is empty, replace surrounds all characters in the input string with the replacement.
+				perReplaceWithSz.Max = sz.Max + 1
+				// and the longest possible original string length is retained
+				retainedSz.Max = sz.Max
+			} else {
+				// largest input size composed of the smallest possible substrings being replaced by largest possible replacement
+				// as many times as possible
+				perReplaceWithSz.Max = uint64(math.Ceil(float64(sz.Max) / float64(toReplaceSz.Min)))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:139]: doesn't the max depend on whether replaceWithSz.Max is greater than toReplaceSz.Max or not?

if `replaceWithSz.Max > toReplaceSz.Max` (e.g. replace "a" with "aa"), the max result size occurs if we replace as many times as possible (`ceil(sz.Max/toReplaceSz.Min)`, as you have here)

if `replaceWithSz.Max <= toReplaceSz.Max` (e.g. replace "aa" with "a"), the max result size occurs if we replace zero times, right? Shouldn't that be represented here somewhere, and just use sz.Max as the result max?

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 2]
```diff
@@ -126,13 +126,32 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
 			sz := l.sizeEstimate(*target)
 			toReplaceSz := l.sizeEstimate(args[0])
 			replaceWithSz := l.sizeEstimate(args[1])
-			// smallest possible result: smallest input size composed of the largest possible substrings being replaced by smallest possible replacement
-			minSz := uint64(math.Ceil(float64(sz.Min)/float64(toReplaceSz.Max))) * replaceWithSz.Min
-			// largest possible result: largest input size composed of the smallest possible substrings being replaced by largest possible replacement
-			maxSz := uint64(math.Ceil(float64(sz.Max)/float64(toReplaceSz.Min))) * replaceWithSz.Max
+
+			var perReplaceWithSz, retainedSz checker.SizeEstimate
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:130]: this name is unclear - is `perReplaceWithSz` the min/max of a single replace, or the min/max number of times we can replace? if the latter, maybe `replaceCount` would be clearer

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 3]
```diff
@@ -126,13 +126,32 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
...
+			}
+			if toReplaceSz.Max == 0 {
+				// if the string being replaced is empty, replace surrounds all characters in the input string with the replacement.
+				perReplaceWithSz.Min = sz.Min + 1
+				// and the shortest possible original string length is retained
+				retainedSz.Min = sz.Min
+			} else {
+				// smallest input size composed of the largest possible substrings being replaced by smallest possible replacement
+				// as many times as possible
+				perReplaceWithSz.Min = uint64(math.Ceil(float64(sz.Min) / float64(toReplaceSz.Max)))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:149]: same comment about min in reverse... I think this depends on whether replaceWithSz.Min is greater than toReplaceSz.Min

if `replaceWithSz.Min < toReplaceSz.Min` (e.g. replace "aa" with "a"), the min result size occurs if we replace as much as possible as many times as possible (`ceil(sz.Min/toReplaceSz.Max)`, as you have here)

if `replaceWithSz.Min >= toReplaceSz.Min (e.g. replace "a" with "aa"), the max result size occurs if we replace zero times, right? Shouldn't that be represented here somewhere, and just use sz.Min as the result min?

**Comment by liggitt** [comment]
/hold
I think the min/max is not quite correct

**Comment by liggitt** [comment]
> https://gist.github.com/jpbetz/032b54ce4ef8f5427c4812c023f6c693 shows the exact difference, which is that `uint64(+Inf)` results in 9223372036854775808 for amd64 but 18446744073709551615 for ppc64le (and arm64?).
> 
> The fix is to avoid the divide the creates the +Inf by handling the cost edge cases for `replace()`. I've verified this fix works on both architectures.

Nice work tracking that down (and many thanks to @Rajalakshmi-Girish and the folks running the ppc tests for raising the issue right away). Do know if we have other Inf+ operations lurking in our cost estimation?

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 4]
```diff
@@ -126,13 +126,43 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
...
-			// smallest possible result: smallest input size composed of the largest possible substrings being replaced by smallest possible replacement
-			minSz := uint64(math.Ceil(float64(sz.Min)/float64(toReplaceSz.Max))) * replaceWithSz.Min
-			// largest possible result: largest input size composed of the smallest possible substrings being replaced by largest possible replacement
-			maxSz := uint64(math.Ceil(float64(sz.Max)/float64(toReplaceSz.Min))) * replaceWithSz.Max
+
+			var replaceCount, retainedSz checker.SizeEstimate
+			// find the longest replacement:
+			if toReplaceSz.Min == 0 {
+				// if the string being replaced is empty, replace surrounds all characters in the input string with the replacement.
+				replaceCount.Max = sz.Max + 1
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:134]: sz.Max will never be MAX_INT and cause this to overflow/wrap, right?

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/cel/library/cost.go]: Thx, I clearly was working too late last night :)  Fixed!

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go (hunk 4 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/cel/library/cost.go]: Added a check. The rest of the match is using the utils that check for overflow/wrap so we should be good now.

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 5]
```diff
@@ -126,13 +126,47 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
...
-			// largest possible result: largest input size composed of the smallest possible substrings being replaced by largest possible replacement
-			maxSz := uint64(math.Ceil(float64(sz.Max)/float64(toReplaceSz.Min))) * replaceWithSz.Max
+
+			var replaceCount, retainedSz checker.SizeEstimate
+			// find the longest replacement:
+			if toReplaceSz.Min == 0 {
+				// if the string being replaced is empty, replace surrounds all characters in the input string with the replacement.
+				if replaceCount.Max < math.MaxUint64 {
+					replaceCount.Max = sz.Max + 1
+				}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:136]:
```suggestion
				if sz.Max < math.MaxUint64 {
					replaceCount.Max = sz.Max + 1
				} else {
					replaceCount.Max = sz.Max
				}
```

File: staging/src/k8s.io/apiserver/pkg/cel/library/cost.go [hunk 6]
```diff
@@ -126,13 +126,47 @@ func (l *CostEstimator) EstimateCallCost(function, overloadId string, target *ch
...
+				// as many times as possible.
+				replaceCount.Max = uint64(math.Ceil(float64(sz.Max) / float64(toReplaceSz.Min)))
+			}
+
+			// find the shortest replacement:
+			if toReplaceSz.Max == 0 {
+				// if the string being replaced is empty, replace surrounds all characters in the input string with the replacement.
+				if replaceCount.Min < math.MaxUint64 {
+					replaceCount.Min = sz.Min + 1
+				}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/cel/library/cost.go:154]:
```suggestion
				if sz.Min < math.MaxUint64 {
					replaceCount.Min = sz.Min + 1
				} else {
					replaceCount.Min = sz.Min
				}
```

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/cel/library/cost.go]: Thx!

**Comment by liggitt** [comment]
once merged, update the cherry-pick PRs (https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+-base%3Amaster+119800) to include this commit

**Comment by BenTheElder** [comment]
When you see a flaking unit test, please file an issue or add a reference to the existing issue, unit tests should NOT be flaking : (

I don't see one so I filed https://github.com/kubernetes/kubernetes/issues/120112

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Fix CEL estimated cost of replace() to handle a zero length replacement string correctly.
-Previously this would cause the estimated cost to be higher than it should be.
+Fixed CEL estimated cost of `replace()` to handle a zero length replacement string correctly.
+Previously this would cause the estimated cost to be higher than it should be.
```

**Comment by jpbetz** [comment]
> Changelog suggestion

Applied. Thanks!

**Pull request merged**